astacus/coordinator/plugins/clickhouse/steps.py

```diff
@@ -191,36 +191,42 @@ async def revoke_write_on_table(self, table: Table, user_name: bytes) -> None:
             f"REVOKE INSERT, ALTER UPDATE, ALTER DELETE ON {table.escaped_sql_identifier} FROM {escaped_user_name}"
         )
         await asyncio.gather(*(client.execute(revoke_statement.encode()) for client in self.clients))
-        await self.wait_for_access_type_grant(user_name=user_name, table=table, expected_count=0)
+        await self.wait_for_access_type_grant(user_name=user_name, table=table, expected_count=0, filter_partial_revoke=True)
```

Comment:
Can this return if the user has a grant on the database but before the revoke goes through?

Reply:
If I comment out

# await asyncio.gather(*(client.execute(revoke_statement.encode()) for client in self.clients))

this line still passes in the test `test_keeper_map_table_read_only_step`.

Would it make more sense to actually just check _for_ the partial revoke.  Rather than filtering out those rows.

Reply:
```python
sql = """
SELECT access_type
FROM system.grants
WHERE ...
AND is_partial_revoke = 1
"""
revokes = (await client.execute(sql)).data
for type in ('INSERT', 'ALTER UPDATE', 'ALTER DELETE'):
    assert type in revokes
```

Reply:
Well not really either. Here is the result for `SHOW GRANTS` after the execution of the grant altering statement during ``tests.integration.coordinator.plugins.clickhouse.test_steps.test_keeper_map_table_read_only_step``

```
INFO     astacus.coordinator.plugins.clickhouse.steps:steps.py:227 Grants for user b'foobar': [['GRANT SHOW, SELECT, INSERT, ALTER UPDATE, ALTER DELETE, ALTER COLUMN, ALTER MODIFY COMMENT, ALTER INDEX, ALTER PROJECTION, ALTER CONSTRAINT, ALTER TTL, ALTER SETTINGS, ALTER VIEW, CREATE TABLE, DROP TABLE, TRUNCATE, OPTIMIZE, SYSTEM SYNC REPLICA ON keeperdata.* TO foobar WITH GRANT OPTION'], ['REVOKE INSERT, ALTER UPDATE, ALTER DELETE ON keeperdata.keepertable FROM foobar']]
INFO     astacus.coordinator.plugins.clickhouse.steps:steps.py:227 Grants for user b'alice': [['GRANT SELECT ON default.* TO alice'], ['GRANT bob TO alice']]
INFO     astacus.coordinator.plugins.clickhouse.steps:steps.py:227 Grants for user b'z_\x80_enjoyer': []
INFO     astacus.coordinator.plugins.clickhouse.steps:steps.py:227 Grants for user b'foobar': [['GRANT SHOW, SELECT, INSERT, ALTER UPDATE, ALTER DELETE, ALTER COLUMN, ALTER MODIFY COMMENT, ALTER INDEX, ALTER PROJECTION, ALTER CONSTRAINT, ALTER TTL, ALTER SETTINGS, ALTER VIEW, CREATE TABLE, DROP TABLE, TRUNCATE, OPTIMIZE, SYSTEM SYNC REPLICA ON keeperdata.* TO foobar WITH GRANT OPTION'], ['REVOKE GRANT OPTION FOR INSERT, ALTER UPDATE, ALTER DELETE ON keeperdata.keepertable FROM foobar']]
INFO     astacus.coordinator.plugins.clickhouse.steps:steps.py:227 Grants for user b'alice': [['GRANT SELECT ON default.* TO alice'], ['GRANT INSERT, ALTER UPDATE, ALTER DELETE ON keeperdata.keepertable TO alice'], ['GRANT bob TO alice']]
INFO     astacus.coordinator.plugins.clickhouse.steps:steps.py:227 Grants for user b'z_\x80_enjoyer': [['GRANT INSERT, ALTER UPDATE, ALTER DELETE ON keeperdata.keepertable TO `z_�_enjoyer`']]
```

Reply:
Yea okay I see.  Feels like it should be an upstream feature.

Reply:
But that's the catch, `is_partial_revoke` isn't a good discriminating variable: if you filter by it, and you have the grants as below, you might return before the grants are revoked (and those rows are gone).
```

user_name         role_name    access_type    database    table       column      is_partial_revoke    grant_option
----------------  -----------  -------------  ----------  ----------  --------  -------------------  --------------
aiven_monitoring               INSERT         default     keeper_map                              0               0
aiven_monitoring               ALTER UPDATE   default     keeper_map                              0               0
aiven_monitoring               ALTER DELETE   default     keeper_map                              0               0
alice                          INSERT         default     keeper_map                              0               0
alice                          ALTER UPDATE   default     keeper_map                              0               0
alice                          ALTER DELETE   default     keeper_map                              0               0
bob                            INSERT         default     keeper_map                              1               1
bob                            ALTER UPDATE   default     keeper_map                              1               1
bob                            ALTER DELETE   default     keeper_map                              1               1
```